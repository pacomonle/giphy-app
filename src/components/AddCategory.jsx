import React, { useState } from 'react'
import PropTypes from 'prop-types';

const AddCategory = ({setCategories}) => {

const [inputValue, setInputValue] = useState('')

const handleSubmit = (e) => {
    e.preventDefault();
    console.log('submit hecho')
    if(inputValue.trim().length > 2){
        setCategories(cat => [inputValue, ...cat])
        setInputValue('')
    }   
}

const handleChange = (e) => {
    console.log('handleChange activado', inputValue)
    setInputValue(e.target.value)
}
    return (
        <div>
            <h2>AddCategory</h2>
            <form onSubmit={handleSubmit}>
                    <input
                    type="text" 
                    placeholder='selecciona categoria'
                    value={inputValue}
                    onChange={(e) => {handleChange(e)}}
                    />
            </form>
            
        </div>
    )
}


AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}

export default AddCategory
