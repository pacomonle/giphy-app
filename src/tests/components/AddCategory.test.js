import { shallow } from 'enzyme';
import '@testing-library/jest-dom';
import React from 'react';
import AddCategory from '../../components/AddCategory';



describe('Pruebas en <AddCategory></AddCategory>', () => {

   // const setCategories = () => {}
   const setCategories = jest.fn()
   let wraper = shallow(<AddCategory setCategories={setCategories}></AddCategory>)
     // ciclo de vida de las pruebas -> beforEach metodo que se ejecutan al iniciar cualquier prueba
   beforeEach(()=>{
       console.log('beforEach!!!!')
       jest.clearAllMocks
       wraper = shallow(<AddCategory setCategories={setCategories}></AddCategory>)
   })

    test('should mostrarse correctamente', () => {
        expect(wraper).toMatchSnapshot()
    })

    test('should cambiar el input o caja texto', () => {
        // 1. seleccionar el input
        const input = wraper.find('input')
        const value = 'Hola Mundo'
        // 2. simular el evento - en la simulacion los eventos no llevan el on
        /*hay que mandar el evento e.target.value*/
        input.simulate('change', {target:{value}})
        
        // comprobar si el setInputValue actualiza el estado, hay que crear <p>{inputValue}</p>
        //  expect(wraper.find('p').trim()).toBe(value)
        
    })

    test('should no postear con el submit', () => {
        // 1. seleccionar y simular el submit
        const submit = wraper.find('form').simulate('submit', {preventDefault(){}})
        // 2. como no se cumple la condicion comprobar que setCategories no es llamado
        expect(setCategories).not.toHaveBeenCalled()
    })

    
    test('should llamar al setCategories y limpiar la caja de texto', () => {
        const value='Hola MUndo'
        // 1. simular input change
        const change = wraper.find('input').simulate('change', {target: {value}})
        // 2. simular el submit
        const submit = wraper.find('form').simulate('submit', {preventDefault(){}})
        // 3. comprobar que el setCategories es llamado
        expect(setCategories).toHaveBeenCalled()
        expect(setCategories).toHaveBeenCalledTimes(1)
        expect(setCategories).toHaveBeenCalledWith(expect.any(Function))
        // 4 comprobar al salir del submit la caja del input se vacia
        console.log('ver', wraper.find('input').props())
        expect(wraper.find('input').prop('value')).toBe('')
    })
    
    
    
})
