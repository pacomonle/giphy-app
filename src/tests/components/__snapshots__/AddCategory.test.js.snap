// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddCategory></AddCategory> should mostrarse correctamente 1`] = `
<div>
  <h2>
    AddCategory
  </h2>
  <form
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="selecciona categoria"
      type="text"
      value=""
    />
  </form>
</div>
`;
